modules/040-node-manager/hooks/fencing_controller.go

```diff
@@ -0,0 +1,188 @@
+	"github.com/flant/shell-operator/pkg/kube/object_patch"
+	"github.com/flant/shell-operator/pkg/kube_events_manager/types"
+	coordinationv1 "k8s.io/api/coordination/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/fields"
+
+	"github.com/deckhouse/deckhouse/go_lib/dependency"
+)
+
+const (
+	nodesSnapshot            = "nodes"
+	leasesSnapshot           = "leases"
+	podsSnapshot             = "pods"
+	fencingControllerTimeout = time.Duration(60) * time.Second
+)
+
+var maintanenceAnnotations = []string{
+	`update.node.deckhouse.io/disruption-approved`,
+	`update.node.deckhouse.io/approved`,
+	`node-manager.deckhouse.io/fencing-disable`,
+}
+
+var _ = sdk.RegisterFunc(&go_hook.HookConfig{
+	Queue: "/modules/node-manager/fecning",
+	Kubernetes: []go_hook.KubernetesConfig{
+		{
+			Name:       nodesSnapshot,
+			ApiVersion: "v1",
+			Kind:       "Node",
+			LabelSelector: &metav1.LabelSelector{
+				MatchExpressions: []metav1.LabelSelectorRequirement{
+					{
+						Key:      "node-manager.deckhouse.io/fencing-enabled",
+						Operator: metav1.LabelSelectorOpExists,
+					},
+				},
+			},
+			FilterFunc: fencingControllerNodeFilter,
+		},
+		{
+			Name:       leasesSnapshot,
+			ApiVersion: "coordination.k8s.io/v1",
+			Kind:       "Lease",
+			NamespaceSelector: &types.NamespaceSelector{
+				NameSelector: &types.NameSelector{
+					MatchNames: []string{"kube-node-lease"},
+				},
+			},
+			FilterFunc: fencingControllerLeaseFilter,
+		},
+	},
+}, dependency.WithExternalDependencies(fencingControllerHandler))
+
+type fencingControllerNodeResult struct {
+	Name          string
+	NodeGroupName string
+}
+
+type fencingControllerLeaseResult struct {
+	NodeName  string
+	RenewTime time.Time
+}
+
+func fencingControllerNodeFilter(obj *unstructured.Unstructured) (go_hook.FilterResult, error) {
+	var res fencingControllerNodeResult
+
+	for _, annotation := range maintanenceAnnotations {
+		_, annotationExists := obj.GetAnnotations()[annotation]
+		if annotationExists {
+			return nil, nil
+		}
+	}
+
+	res.Name = obj.GetName()
+	return res, nil
+}
+
+func fencingControllerLeaseFilter(obj *unstructured.Unstructured) (go_hook.FilterResult, error) {
+	var lease coordinationv1.Lease
+	err := sdk.FromUnstructured(obj, &lease)
+	if err != nil {
+		return nil, err
+	}
+	return fencingControllerLeaseResult{
+		NodeName:  *lease.Spec.HolderIdentity,
+		RenewTime: lease.Spec.RenewTime.Time,
+	}, nil
+}
+
+func fencingControllerHandler(input *go_hook.HookInput, dc dependency.Container) error {
+	if len(input.Snapshots[nodesSnapshot]) == 0 {
+		// No nodes with enabled fencing -> nothing to do
+		return nil
+	}
+
+	// make map with nodes
+	nodesMap := make(map[string]struct{})
+	for _, nodeRaw := range input.Snapshots[nodesSnapshot] {
+		if nodeRaw != nil {
+			node := nodeRaw.(fencingControllerNodeResult)
+			nodesMap[node.Name] = struct{}{}
+		}
+	}
```

Comment:
I think we can use set.Set from go_lib here

Reply:
fixed